### PR TITLE
Anonymous login data and dedicated server support (WIP)

### PR DESCRIPTION
**Pull request by bubby932**
* Adds anonymous analytics for monitoring login times & density.
* Adds WIP support for dedicated servers.
* Adds a number of endpoints for the Room Management API. (see my review)

### CODE REVIEW

**Comment by bubby932** [comment]
the robot

File: routers/auth.js [hunk 1]
```diff
@@ -21,6 +21,32 @@ const accountCreationLimit = rateLimit({
     'standardHeaders': true
 });
 
+router.get('/photon-info', async (req, res) => {
```
Comment (bubby932) [inline routers/auth.js:24]: Endpoint for dedicated server support, not complete yet.

File: routers/auth.js [hunk 2]
```diff
@@ -158,6 +184,12 @@ router.post("/login", async (req, res) => {
             await helpers.PushPlayerData(userID, data);
         }
 
+        const mongo = require('../index').mongoClient;
```
Comment (bubby932) [inline routers/auth.js:187]: Login analytics

File: routers/auth.js [hunk 3]
```diff
@@ -174,6 +206,14 @@ router.post("/login", async (req, res) => {
             data.auth.logins.push(attempt);
             await helpers.PushPlayerData(userID, data);
         }
+
```
Comment (bubby932) [inline routers/auth.js:209]: More analytics

**Review by bubby932 — COMMENTED** [review]
explained

**Comment by bubby932** [comment]
@Bobrobot1 :politehalo:

File: administrator_dashboard.js [hunk 4]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-await-in-loop */
```
Comment (bubby932) [inline administrator_dashboard.js:1]: this is a non-essential file so we can be lax

File: package.json [hunk 5]
```diff
@@ -18,19 +18,19 @@
     "dotenv": "^16.0.3",
     "express": "^4.18.2",
     "express-fileupload": "^1.2.1",
-    "express-rate-limit": "^6.6.0",
+    "express-rate-limit": "^6.7.0",
```
Comment (bubby932) [inline package.json:21]: This file is all @dependabot bumps.

File: routers/rooms.js [hunk 6]
```diff
@@ -64,7 +64,7 @@ router.route("/room/:room_id/info")
             }
 
             // refetch the room using projection so we're not exposing permissions and subroom data to the user
-            const room_visible = await room_collection.findOne({_id: {$eq: room_id, $exists: true}}, {projection: {_id: 1, name: 1, description: 1, creator_id: 1, tags: 1, created_at: 1, visits: 1, homeSubroomId: 1, cover_image_id: 1}});
+            const room_visible = await room_collection.findOne({_id: {$eq: room_id, $exists: true}}, {projection: {_id: 1, name: 1, description: 1, creator_id: 1, tags: 1, created_at: 1, visits: 1, homeSubroomId: 1, cover_image_id: 1, contentFlags: 1}});
```
Comment (bubby932) [inline routers/rooms.js:67]: Added support for content flags in projection.

File: routers/rooms.js [hunk 7]
```diff
@@ -151,7 +151,8 @@ router.get("/search", authenticateToken_optional, async (req, res) => {
             tags: item.tags,
             visits: item.visits,
             created_at: item.created_at,
-            cover_image_id: item.cover_image_id
+            cover_image_id: item.cover_image_id,
```
Comment (bubby932) [inline routers/rooms.js:154]: Added support for content flags in search.

File: routers/rooms.js [hunk 8]
```diff
@@ -404,6 +405,174 @@ router.post('/room/:id/subrooms/:subroom_id/versions/public', authenticateToken,
     }
 });
 
+router.post('/room/:id/tags', authenticateToken, requiresRoomPermission("manageTags"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:408]: Room management API - Room Set Tags

File: routers/rooms.js [hunk 9]
```diff
@@ -404,6 +405,174 @@ router.post('/room/:id/subrooms/:subroom_id/versions/public', authenticateToken,
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal error occurred and we could not process your request."
+        });
+        throw ex;
+    }
+});
+
+router.post('/room/:id/content_flags', authenticateToken, requiresRoomPermission("manageTags"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:455]: Room management API - Room Set Content Flags

File: routers/rooms.js [hunk 10]
```diff
@@ -404,6 +405,174 @@ router.post('/room/:id/subrooms/:subroom_id/versions/public', authenticateToken,
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal error occurred and we could not process your request."
+        });
+        throw ex;
+    }
+});
+
+router.post('/room/:id/description', authenticateToken, requiresRoomPermission("editDescription"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:502]: Room management API - Room Set Description

File: routers/rooms.js [hunk 11]
```diff
@@ -404,6 +405,174 @@ router.post('/room/:id/subrooms/:subroom_id/versions/public', authenticateToken,
...
+    } catch (ex) {
+        res.status(500).json({
+            code: "internal_error",
+            message: "An internal error occurred and we could not process your request."
+        });
+        throw ex;
+    }
+});
+
+router.get('/room/:id/subrooms/list', authenticateToken, requiresRoomPermission("manageSubrooms"), async (req, res) => {
```
Comment (bubby932) [inline routers/rooms.js:542]: Room management API - Room Get Subrooms

File: routers/rooms.js [hunk 12]
```diff
@@ -592,6 +761,45 @@ async function canViewRoom(req, res, next) {
     next();
 }
 
+function requiresRoomPermission(permission) {
```
Comment (bubby932) [inline routers/rooms.js:764]: New helper function for permissions, similar to authenticateTokenAndTag(string)

**Review by bubby932 — COMMENTED** [review]
Explained everything again

File: .eslintrc.json [hunk 13]
```diff
@@ -16,7 +16,6 @@
         "no-invalid-this": "error",
         "semi-style": "error",
         "semi-spacing": "error",
-        "no-useless-escape": "off",
-        "no-await-in-loop": "error"
```
Comment (bubby932) [inline .eslintrc.json:20]: gone

File: routers/img.js [hunk 14]
```diff
@@ -48,7 +48,8 @@ const imageMetadataTemplate = {
         tags: [
             'photo'
         ]
-    }
+    },
+    visibility: "public"
```
Comment (bubby932) [inline routers/img.js:52]: unlisted photo support

File: routers/img.js [hunk 15]
```diff
@@ -107,6 +108,8 @@ router.post("/upload", uploadRateLimit, middleware.authenticateToken, async (req
 
         MetaData.social.tags = JSON.parse(tags);
 
+        if (req.query.visibility == "unlisted") MetaData.visibility = "unlisted";
```
Comment (bubby932) [inline routers/img.js:111]: unlisted photo support

File: routers/rooms.js [hunk 16]
```diff
@@ -372,16 +373,16 @@ router.post('/room/:id/subrooms/:subroom_id/versions/public', authenticateToken,
...
             "message": "There is no version of this room associated with that ID."
         });
 
-        const str = `subrooms.${id}.publicVersionId`;
+        const str = `subrooms.${subroom_id}.publicVersionId`;
 
         // this is a painful workaround to fix javascript's assfuckery -Rose
         var setFilter = {};
-        setFilter[str] = id;
+        setFilter[str] = version_id;
```
Comment (bubby932) [inline routers/rooms.js:385]: fixed some issues setting public version

File: routers/social.js [hunk 17]
```diff
@@ -16,7 +16,14 @@ router.get("/imgfeed", async (req, res) => {
         const db = client.db(process.env.MONGOOSE_DATABASE_NAME);
         const image_collection = db.collection("images");
           
-        const all_images = await image_collection.find().toArray();
+        const all_images = await image_collection.find(
```
Comment (bubby932) [inline routers/social.js:19]: only show public images

**Review by bubby932 — COMMENTED** [review]
explained more things

**Review by Bobrobot1 — APPROVED** [review]

**Pull request merged**